src/Grpc.AspNetCore.Server/Internal/GrpcProtocolHelpers.cs

```diff
@@ -163,5 +171,63 @@ public static void AppendStatusTrailers(HttpResponse response, Status status)
             response.AppendTrailer(GrpcProtocolConstants.StatusTrailer, status.StatusCode.ToTrailerString());
             response.AppendTrailer(GrpcProtocolConstants.MessageTrailer, status.Detail);
         }
+
+        public static AuthContext CreateAuthContext(X509Certificate2 clientCertificate)
+        {
+            // Map X509Certificate2 values to AuthContext. The name/values come BoringSSL via C Core
+            // https://github.com/grpc/grpc/blob/a3cc5361e6f6eb679ccf5c36ecc6d0ca41b64f4f/src/core/lib/security/security_connector/ssl_utils.cc#L206-L248
+
+            var properties = new Dictionary<string, List<AuthProperty>>();
+
+            string peerIdentityPropertyName = null;
+
+            foreach (var extension in clientCertificate.Extensions)
+            {
+                if (extension.Oid.Value == GrpcProtocolConstants.X509SubjectAlternativeNameId)
+                {
+                    var formattedAlternativeNamesText = extension.Format(multiLine: true);
+
+                    // The format of subject names changes between operating systems
+                    // Parse subject names using a regex that supports all major operating systems (Windows/Mac/Linux)
+                    foreach (Match match in SubjectAlternativeNameRegex.Matches(formattedAlternativeNamesText))
+                    {
+                        if (match.Success)
+                        {
+                            var domain = match.Groups["domain"].Value;
+
+                            AddProperty(properties, GrpcProtocolConstants.X509SubjectAlternativeNameKey, domain);
+
+                            if (peerIdentityPropertyName == null)
+                            {
+                                peerIdentityPropertyName = GrpcProtocolConstants.X509SubjectAlternativeNameKey;
+                            }
+                        }
+                    }
+                }
+            }
+
+            var commonName = clientCertificate.GetNameInfo(X509NameType.SimpleName, false);
```

Comment:
X509 standards state if you have SAN strings (the alternative names above) then the subject/common name is invalid. You are still parsing it, if a SAN exists. This should only run if there if there is no SAN string.

Also don't use GetNameInfo for this, there's a SubjectName property.

Reply:
To be compatible with other implementations of AuthContext the common name property should be in the collection. You can see on line 215 that the common name will only be used as the peer identity if there is no SAN.